R/readKeyedData.R

```diff
@@ -0,0 +1,203 @@
+  
+  #### Trim key to only read in matched vars ####
+  if(verbose) {print(paste('dropping', sum(is.na(key.df$var)), 'unpaired headers.'))}
+  key.df <- key.df %>%
+    filter(!is.na(var))
+  
+  ####Expand the regular expressions in the key ####
+  varsToExpand <- key.df %>%
+    filter(softType != 'value' & !is.na(softType), #pull softTypes that are not values
+           #If var doesn't match value vars assume they are regular expressions
+           !var %in% (key.df %>% filter(softType == 'value'))$var) %>%
+    group_by(header) 
+  
+  if(nrow(varsToExpand) > 0){
+    key.df <- varsToExpand %>%
+      do((function(xx=.,
+                   varlist = (key.df %>% filter(!is.na(var) & softType == 'value'))$var){
+        return(data.frame(fullVar=varlist[grepl(xx$var, varlist)],
+                          stringsAsFactors=FALSE))
+      })(.)) %>% #fill in the vars that match the regular expression
+      right_join(key.df, by='header') %>% #merge back into main key
+      group_by(header, dataframe, softType) %>%
+      mutate(var = if_else(is.na(fullVar), var, fullVar)) %>%
+      arrange(var) %>%
+      ungroup() %>%
+      select(-fullVar)
+  }
+  
+  ### Check that all vars are either long or wide and there are wide ids flaged ###
+  assertthat::assert_that(any(is.na(key.df$softType) & !is.na(key.df$flagID)), 
+             msg='A long data type can not be a flaged id')
+  assertthat::assert_that(any(!is.na(key.df$softType) & is.na(key.df$flagID)),
+             msg='There must be a flagged id for the wide data types')
+  
+  
+  wide.df <- NULL
+  long.df <- NULL
+  
+  for(tableIndex in 1:numIterate){
+    if(verbose) print(paste('reading table', tableIndex))
+    ### read in orginal data table ###
+    if(readingCSVfiles){
+      data.df <- read_csv(filename[tableIndex], skip=skipRows, 
+                          col_names= tableIndex != verticalTable)
+    }else{
+      if(readxl::excel_sheets(filename)[tableIndex] %in% excludeSheets){
+        next()
+      }
+      data.df <- readxl::read_excel(filename, sheet=tableIndex, skip=skipRows,
+                                    col_names= tableIndex != verticalTable)
+    }
+    
+    ### flip things around if it's a vertical format ###
+    if(tableIndex == verticalTable){
+      data.df <- bind_rows(data.df, setNames(names(data.df), names(data.df)))
+      data.df <- as.data.frame(t(data.df), stringsAsFactors=FALSE)
+      names(data.df) <- data.df[1,]
+      data.df <- data.df[-1,]
+    }
+    
+    ### drop any preidentified rows and empty columns ###
+    if(all(!is.na(dropRows))){
+      data.df <- data.df[-1*dropRows,] 
+    }
+    
+    ### pull the wide data ###
+    wideKey.df <- key.df %>%
+      filter(is.na(softType), ##define wide tables by absence of soft Type
+             header %in% names(data.df)) ##only pull the headers we care about
+    
+    wideTemp <- data.df %>% 
+      select_(quote(wideKey.df$header)) %>%
+      select_if(~any(!is.na(.))) #drop empty columns
```

Comment:
Whoah funky

Reply:
I know right? I was delighted to stumble across this little gem in my frantic google code searches.